### ISSUE TITLE
feat: VSR AutoResearch — Autonomous Model Improvement Loop on Local GPU

### ISSUE THREAD

**Issue by yossiovadia** [yossiovadia/semantic-router#59]
## Summary

Implement an autonomous experimentation loop (inspired by [karpathy/autoresearch](https://github.com/karpathy/autoresearch)) that continuously improves VSR classification, PII, jailbreak, and LoRA models by running overnight on a single consumer GPU (RTX 4090).

**Agent brain:** Local Ollama (Qwen 3.5 35B MoE) — zero API cost  
**Training compute:** RTX 4090 24GB VRAM  
**GPU sharing:** Time-share — unload LLM during training, reload for experiment planning  

## Motivation

VSR ships 8+ specialized models (`mom-domain-classifier`, `mom-pii-classifier`, `mom-jailbreak-classifier`, `mom-halugate-*`, `mom-embedding-*`, LoRA specialists). Each has hyperparameters, architecture choices, and dataset configurations that affect quality. Currently these are tuned manually. An autonomous loop could run 40-50 experiments overnight, systematically exploring the search space and improving each model.

### Phase 1: Domain Classifier (highest routing impact)
- **Training script:** `src/training/classifier_model_fine_tuning/ft_linear.py`
- **Architecture:** ModernBERT-base → sequence classification (10 MMLU-Pro domains)
- **Eval metric:** Per-domain F1, overall accuracy
- **Search space:** backbone (bert/roberta/modernbert/deberta), learning rate, batch size, weight decay, warmup ratio, epochs, dropout

### Phase 2: Jailbreak Classifier (security)
- **Training script:** `src/training/prompt_guard_fine_tuning/jailbreak_bert_finetuning.py`
- **Architecture:** ModernBERT-base → binary classification
- **Eval metric:** F1 at threshold, precision/recall balance
- **Search space:** Same backbone options, dataset combinations (salad-data, toxic-chat, spml-injection), max-samples-per-source

### Phase 3: PII Classifier (privacy)
- **Training script:** `src/training/pii_model_fine_tuning/pii_bert_finetuning.py`
- **Architecture:** ModernBERT-base → token classification (BIO tagging, 6 entity types)
- **Eval metric:** Per-entity F1, token accuracy
- **Search space:** backbone, dataset (presidio vs ai4privacy), all hyperparams

### Phase 4: LoRA Specialists (per-domain solvers)
- **Training script:** `src/training/training_lora/mmlu_pro_solver_lora/ft_qwen3_mmlu_solver_lora.py`
- **Architecture:** Qwen3-0.6B + LoRA (6 specialist types)
- **Eval metric:** Per-category accuracy (baseline 10% random → target 40-60%)
- **Search space:** LoRA rank/alpha, learning rate, epochs, samples per category

### Config-Driven, Not Code-Editing

Unlike the original autoresearch (agent edits `train.py` directly), we use a **config-driven** approach:
- AI agent edits `experiment.json` (hyperparams, model choice, dataset config)
- `train_wrapper.py` reads the config and calls existing VSR training scripts with appropriate arguments
- This prevents the agent from breaking training code
- Search space is constrained but large enough for meaningful optimization

```json
{
  "phase": "domain-classifier",
  "backbone": "modernbert-base",
  "learning_rate": 2e-5,
  "batch_size": 16,
  "weight_decay": 0.1,
  "warmup_ratio": 0.06,
  "max_epochs": 3,
  "early_stopping_patience": 2,
  "dropout": 0.1,
  "hypothesis": "Testing higher dropout to reduce overfitting seen in exp #12"
}
```

### Experiment Loop

```
┌─────────────────────────────────────────────────┐
│  runner.sh (orchestrator)                       │
│                                                 │
│  1. Load Qwen 3.5 into Ollama (~20s)           │
│  2. Send context: program.md + last 5 results  │
│  3. Ollama outputs new experiment.json          │
│  4. Unload Qwen 3.5 (free GPU for training)    │
│  5. python train_wrapper.py (fixed time budget) │
│  6. Append results to results.tsv               │
│  7. If improved → save checkpoint to best/      │
│  8. Goto 1                                      │
└─────────────────────────────────────────────────┘
```

### Time Budget Per Experiment
- **Classifiers (Phase 1-3):** 10 minutes (ModernBERT-base converges fast on 4090)
- **LoRA specialists (Phase 4):** 15 minutes (Qwen3-0.6B needs slightly more)
- **Expected throughput:** ~4-5 experiments/hour → **40-50 experiments overnight**

### GPU Time-Sharing
- Ollama Qwen 3.5 uses ~24GB VRAM (full GPU)
- Training uses ~2-4GB VRAM
- Cannot coexist → time-share via Ollama keep_alive=0 to unload between experiments
- ~20 second model load/unload overhead per cycle (negligible vs 10-15 min training)

## Project Structure

```
vsr-autoresearch/
├── program.md              # Agent instructions
├── experiment.json          # Current experiment config (agent edits)
├── results.tsv              # Experiment log (append-only)
├── baseline/                # Baseline metrics per model type
├── best/                    # Best model checkpoints
├── runner.sh                # Orchestrator loop
└── train_wrapper.py         # Config reader → VSR training script caller
```

## Safety & Eval

- **Baseline first:** Run each training script with defaults before any experiments
- **Never regress:** Best checkpoint always preserved in `best/`
- **Append-only log:** Full experiment history in `results.tsv`
- **Agent context:** Sees last N results + current best to learn what works

## Notifications

- Telegram notification on new best result
- Morning summary of overnight experiments  
- Alert if training crashes or loop stalls

## Prerequisites

- Python 3.10+, PyTorch + CUDA, transformers, datasets, sentence-transformers
- Ollama with Qwen 3.5 35B MoE model
- Single NVIDIA GPU (tested on RTX 4090)
- VSR training scripts from `src/training/`

## Future Extensions

- [ ] Multi-objective optimization (accuracy vs inference speed Pareto front)
- [ ] Data augmentation experiments (agent generates synthetic training data)
- [ ] Architecture search (custom ModernBERT configurations)
- [ ] K-fold cross-validation
- [ ] Embedding model training (contrastive learning for `mom-embedding-*`)
- [ ] Upstream PR — contribute best hyperparams back to vllm-project/semantic-router

## Labels

enhancement, training, automation

**Comment by yossiovadia** [comment]
## Plan Improvements (v2)

After reviewing the initial plan, here are critical additions:

### 1. No Tool Calling Needed — Simple Text to JSON

Ollama just gets a text prompt and outputs JSON. No complex tool calling:
```
You are an ML researcher optimizing {phase} models.

Last 5 results:
{results_tail}

Current best: F1={best_f1} (exp #{best_id})
Search space: {search_space}

Respond with ONLY a JSON object for the next experiment.
Include a "hypothesis" field explaining your reasoning.
```
If invalid JSON: retry once, then random search fallback.

### 2. Error Recovery
- **Invalid JSON from Ollama:** Retry once, then random search within bounds
- **Training OOM:** Auto-reduce batch size by half, retry
- **Training crash:** Log error, skip experiment, continue loop
- **Ollama load failure:** Wait 30s, retry 3x, then exit with Telegram alert
- **Max 5 consecutive failures** then pause and alert

### 3. Guided Exploration Strategy
Baked into `program.md` so the agent adapts:
- **Exp 1-5:** Baseline + try each backbone once (broad sweep)
- **Exp 6-15:** Focus on winning backbone, sweep lr + batch size
- **Exp 16+:** Fine-tune around best config (small perturbations)

### 4. Validation Consistency
- Fixed random seed for train/val splits across ALL experiments
- Same validation set = fair comparison. Seed is a constant in experiment.json.

### 5. Disk Management
- Only keep top-3 checkpoints per phase (ranked by F1)
- ModernBERT checkpoint ~700MB, LoRA ~50MB
- Auto-delete worst when a new top-3 enters
- Total budget: ~5GB per phase

### 6. Cross-Phase Learning
When starting Phase 2+, include Phase 1 findings in prompt:
> "In Phase 1, modernbert-base outperformed bert-base by 3.5% F1. Consider starting with modernbert-base."

### 7. Notifications via Telegram
runner.sh sends results on:
- New best result (with metrics + hypothesis)
- Morning summary (total experiments, improvement over baseline, top findings)
- Alert if training crashes or loop stalls

### 8. Morning Report
After loop ends, Ollama reads full results.tsv and generates:
- Total experiments, best per phase, top 3 findings
- What worked vs what didn't
- Recommended next steps
- Improvement delta over baseline

### 9. VSR Repo on yos
Clone full VSR repo on yos (not just copy training scripts) — full context for debugging and accessing datasets/models.

### Implementation Order
1. SSH to yos, check dependencies (Python, PyTorch, CUDA, transformers)
2. Clone VSR repo if not present
3. Create `~/vsr-autoresearch/` project structure
4. Write `program.md`, `runner.sh`, `train_wrapper.py`
5. Run baselines for Phase 1 (domain classifier)
6. Test one full Ollama-train-eval cycle
7. Let it run overnight
8. Review morning report, iterate on program.md

**Comment by yossiovadia** [comment]
## v3: Incorporating Existing Work (Issues #1131, #584, #647)

After reviewing @yossiovadia's prior issues on this repo, this plan significantly overlaps with work already done or identified:

### Already Proven: Domain-Aware Embedding Training (#1131)

The domain-adapted embedding POC (`src/training/domain_adapted_embeddings/`) already implements a more sophisticated version of what we proposed:
- **Iterative hard-negative mining** (not just hyperparameter tuning)
- **+71% MRR improvement** on medical domain
- Published model: `llm-semantic-router/mmbert-embed-medical`
- Training code: `prepare_data.py`, `train.py` with triplet loss

**AutoResearch extension:** Apply the same iterative distillation to other domains (finance, legal, coding, science) automatically. The autoresearch loop would:
1. Pick a target domain
2. Prepare domain-specific Q&A data
3. Run iterative hard-negative mining (2-3 iterations)
4. Eval: MRR@5, Recall@5
5. Compare against baseline embedding
6. Move to next domain

### Known Failure Modes as Eval Criteria

**From #584 (Classification Regression):**
- LoRA classifier achieved only 30% accuracy (7/10 categories failing)
- Strong bias toward psychology/business categories
- AutoResearch MUST track per-domain F1 individually and reject experiments that improve overall F1 but tank specific domains
- The psychology bias pattern is a specific regression test

**From #647 (PII Confidence):**
- SSN detection: 0.6478 confidence (below 0.7 threshold)
- Credit card: 0.3679 confidence (way below threshold)  
- Email without prefix: 0.5742 confidence
- AutoResearch for PII should use entity-specific confidence as the metric, not just overall F1
- Specific eval cases from #647 become the regression test suite

### Revised Phase Structure

| Phase | Target | Approach | Based On |
|-------|--------|----------|----------|
| 1 | Domain Classifier | Hyperparameter optimization + per-domain F1 tracking | Fix #584 patterns |
| 2 | PII Classifier | Hyperparams + dataset mixing (presidio + ai4privacy) | Fix #647 weak entities |
| 3 | Jailbreak Classifier | Hyperparams + dataset combinations | New |
| 4 | LoRA Specialists | LoRA rank/alpha/lr optimization | New |
| **5** | **Domain Embeddings** | **Iterative hard-negative mining per domain** | **Extend #1131 to finance, legal, coding** |
| **6** | **Training Data Augmentation** | **Synthetic data generation for weak categories** | **Address #584 bias** |

### Phase 5: Domain Embedding Expansion (the real value)

Using the proven approach from #1131:

```
For each domain in [finance, legal, coding, science, general]:
  1. Source domain-specific Q&A dataset (HuggingFace or synthetic)
  2. Run prepare_data.py with domain dataset
  3. Run iterative hard-negative mining (train.py)
  4. Eval MRR@5 and Recall@5 vs baseline embedding
  5. If improved: publish domain model to HuggingFace
  6. Log results, move to next domain
```

Target: Replicate the +71% MRR improvement from medical across other domains.

### Phase 6: Synthetic Data for Weak Categories

Address the #584 psychology bias by generating targeted training data:
1. Ollama generates synthetic queries for underperforming categories
2. Train classifier on augmented dataset
3. Eval per-domain F1 — specifically check previously-biased categories
4. Iterate on data quality

This is where the local Ollama agent adds unique value: it can generate domain-specific training examples as part of the experiment loop, not just tune hyperparameters.

### Key Insight

The hyperparameter tuning (Phases 1-4) is table stakes. The real contribution is Phases 5-6: extending proven domain embedding distillation to new domains, and using LLM-generated synthetic data to fix known classifier biases. That is upstream-PR-worthy work.
